services/webnn/ort/graph_builder_ort.cc

```diff
@@ -258,6 +259,658 @@ GraphBuilderOrt::GraphBuilderOrt(
+      input_id_to_pre_matmul_transpose_map;
+
+  // A map of all reshape operations in `mojom::GraphInfo` using transpose's
+  // input operand id as the key and this transpose must be contained in
+  // `input_id_to_pre_matmul_transpose_map`.
+  std::map<uint64_t, const mojom::Operation*>
+      input_id_to_pre_matmul_transpose_reshape_map;
+
+  // A map of all dequantizeLinear operations in `mojom::GraphInfo` using
+  // reshape's input operand id as the key and this dequantizeLinear must be
+  // contained in `input_id_to_pre_matmul_transpose__reshapemap`.
+  std::map<uint64_t, const mojom::Operation*>
+      input_id_to_pre_matmul_transpose_reshape_dq_map;
+
+  // A map to record how many times each operand id is used as one
+  // operation's input edge or the graph's output edge.
+  std::map<uint64_t, uint32_t> operand_id_to_use_count_map;
+  for (const auto& pair : graph_info.id_to_operand_map) {
+    operand_id_to_use_count_map[pair.first] = 0;
+  }
+
+  for (uint64_t graph_output_id : graph_info.output_operands) {
+    ++operand_id_to_use_count_map[graph_output_id];
+  }
+
+  // Iterate from the end of operations instead from the beginning, so we
+  // can easily get the total use count of a fusible base operation's output
+  // edge before visiting it.
+  OperationConnectivity operation_connectivity;
+  for (size_t operation_index = graph_info.operations.size();
+       operation_index-- > 0;) {
+    const auto& operation = graph_info.operations[operation_index];
+    RetrieveOperationConnectivity(
+        operation.get(),
+        /*out_operation_connectivity*/ operation_connectivity);
+
+    for (uint64_t input_id : operation_connectivity.input_ids) {
+      ++operand_id_to_use_count_map[input_id];
+    }
+
+    // Try to find dequantizeLinear that can be fused into following matmul
+    // operations.
+    switch (operation->which()) {
+      case mojom::Operation::Tag::kMatmul: {
+        // Map matmul's inputs to operation, so the following algorithm can find
+        // a dequantizeLinear whose output is consumed by a matmul.
+        CHECK_EQ(operation_connectivity.input_ids.size(), 2U);
+        // We needn't check the result of `try_emplace` here, because if the key
+        // `input_id` is already in container, there must be more than 1 output
+        // edges from a predecessor in which case the dequantizeLinear fusion
+        // won't happen.
+        input_id_to_matmul_map.try_emplace(operation_connectivity.input_ids[1],
+                                           operation.get());
+        break;
+      }
+      case mojom::Operation::Tag::kTranspose: {
+        // If a transpose's output is solely used by a matmul and
+        // transpose's output must be the 2nd input of a matmul.
+        CHECK_EQ(operation_connectivity.output_ids.size(), 1U);
+        uint64_t output_id = operation_connectivity.output_ids[0];
+        // MatMul has only 1 transpose input and the tranpose must have 1 output
```

Comment:
Be consistent and fix typo
```suggestion
        // The matmul has only 1 transpose input and the transpose must have 1 output
```